#Applied Machine Learning Course - Module 2 Notebook
- Matthew Block  
- 31 October 2025  
a brief introduction that describes the problem and the dataset.

In [1]:
# all imports get moved to the top - import each only once

import seaborn as sns
import pandas as pd

## 1. Import & Inspect Data

In [7]:
# Load Titanic dataset
titanic = sns.load_dataset('titanic')

titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [8]:
print(titanic.head(10))
print(titanic.isnull().sum())
print(titanic.describe())

   survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0         0       3    male  22.0      1      0   7.2500        S   Third   
1         1       1  female  38.0      1      0  71.2833        C   First   
2         1       3  female  26.0      0      0   7.9250        S   Third   
3         1       1  female  35.0      1      0  53.1000        S   First   
4         0       3    male  35.0      0      0   8.0500        S   Third   
5         0       3    male   NaN      0      0   8.4583        Q   Third   
6         0       1    male  54.0      0      0  51.8625        S   First   
7         0       3    male   2.0      3      1  21.0750        S   Third   
8         1       3  female  27.0      0      2  11.1333        S   Third   
9         1       2  female  14.0      1      0  30.0708        C  Second   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes

In [9]:
print(titanic.corr(numeric_only=True))

            survived    pclass       age     sibsp     parch      fare  \
survived    1.000000 -0.338481 -0.077221 -0.035322  0.081629  0.257307   
pclass     -0.338481  1.000000 -0.369226  0.083081  0.018443 -0.549500   
age        -0.077221 -0.369226  1.000000 -0.308247 -0.189119  0.096067   
sibsp      -0.035322  0.083081 -0.308247  1.000000  0.414838  0.159651   
parch       0.081629  0.018443 -0.189119  0.414838  1.000000  0.216225   
fare        0.257307 -0.549500  0.096067  0.159651  0.216225  1.000000   
adult_male -0.557080  0.094035  0.280328 -0.253586 -0.349943 -0.182024   
alone      -0.203367  0.135207  0.198270 -0.584471 -0.583398 -0.271832   

            adult_male     alone  
survived     -0.557080 -0.203367  
pclass        0.094035  0.135207  
age           0.280328  0.198270  
sibsp        -0.253586 -0.584471  
parch        -0.349943 -0.583398  
fare         -0.182024 -0.271832  
adult_male    1.000000  0.404744  
alone         0.404744  1.000000  


In [14]:
# Calculate the correlation matrix
corr_matrix = titanic.corr(numeric_only=True)

# Unstack the matrix to get a series of correlation pairs
corr_pairs = corr_matrix.stack()

# Remove self-correlations (e.g., survived vs. survived)
corr_pairs = corr_pairs[corr_pairs.index.get_level_values(0) != corr_pairs.index.get_level_values(1)]

# Create a DataFrame from the series and sort it
sorted_corr = corr_pairs.to_frame('correlation').sort_values('correlation', ascending=False)

# The list contains duplicates (e.g., parch/sibsp and sibsp/parch), so we'll drop them
# by taking every other row.
unique_sorted_corr = sorted_corr.iloc[::2]

print("Sorted Feature Correlations:")
print(unique_sorted_corr)


Sorted Feature Correlations:
                       correlation
sibsp      parch          0.414838
adult_male alone          0.404744
           age            0.280328
survived   fare           0.257307
fare       parch          0.216225
age        alone          0.198270
fare       sibsp          0.159651
pclass     alone          0.135207
age        fare           0.096067
pclass     adult_male     0.094035
sibsp      pclass         0.083081
parch      survived       0.081629
           pclass         0.018443
sibsp      survived      -0.035322
age        survived      -0.077221
fare       adult_male    -0.182024
age        parch         -0.189119
survived   alone         -0.203367
adult_male sibsp         -0.253586
alone      fare          -0.271832
age        sibsp         -0.308247
pclass     survived      -0.338481
adult_male parch         -0.349943
age        pclass        -0.369226
pclass     fare          -0.549500
adult_male survived      -0.557080
parch      alone         -

In [11]:
# Calculate and display the minimum, median, and maximum for 'age'
age_stats = titanic['age'].agg(['min', 'median', 'max'])

print("Age Statistics:")
print(age_stats)


Age Statistics:
min        0.42
median    28.00
max       80.00
Name: age, dtype: float64


### Reflection 1
1) How many data instances are there?
   1) 891
2) How many features are there?
   1) 15
3) What are the names?
   1) survived         0
   2) pclass           0
   3) sex              0
   4) age            177
   5) sibsp            0
   6) parch            0
   7) fare             0
   8) embarked         2
   9) class            0
   10) who              0
   11) adult_male       0
   12) deck           688
   13) embark_town      2
   14) alive            0
   15) alone            0
4) Are there any missing values?
   1) Yes, numbers of missing values are identified in line with Question 3 answers immediately above
5) Are there any non-numeric features?
   1) Yes
6) Are the data instances sorted on any of the attributes?
   1) Appears to be NO
7) What are the units of age?
   1) Looks like years
8) What are the minimum, median and max age?
   1) min        0.42
   2) median    28.00
   3) max       80.00
9)  What two different features have the highest correlation?
    1)  The strongest correlation is a negative correlation between "alone" and "sibsp"
10)  Are there any categorical features that might be useful for prediction?
     1)  Knowing a bit about how people got stuck on or got off the Titanic, both class and deck might be useful in predicting survival.